include/mlir/Dialect/SPIRV/SPIRVBase.td

```diff
@@ -683,22 +697,39 @@ class SPV_Op<string mnemonic, list<OpTrait> traits = []> :
   bit autogenSerialization = 1;
 }
 
-class SPV_ArithmeticOp<string mnemonic, Type type,
-                       list<OpTrait> traits = []> :
-      SPV_Op<mnemonic,
-             !listconcat(traits, [NoSideEffect, SameOperandsAndResultType])> {
+class SPV_BinaryOp<string mnemonic, Type resultType, Type operandsType,
+                   list<OpTrait> traits = []> :
+      SPV_Op<mnemonic, traits> {
   let arguments = (ins
-    SPV_ScalarOrVectorOf<type>:$operand1,
-    SPV_ScalarOrVectorOf<type>:$operand2
+    SPV_ScalarOrVectorOf<operandsType>:$operand1,
+    SPV_ScalarOrVectorOf<operandsType>:$operand2
   );
   let results = (outs
-    SPV_ScalarOrVectorOf<type>:$result
+    SPV_ScalarOrVectorOf<resultType>:$result
   );
   let parser = [{ return impl::parseBinaryOp(parser, result); }];
   let printer = [{ return impl::printBinaryOp(getOperation(), p); }];
   // No additional verification needed in addition to the ODS-generated ones.
   let verifier = [{ return success(); }];
 }
 
+class SPV_ArithmeticOp<string mnemonic, Type type,
+                       list<OpTrait> traits = []> :
+      // Operands type same as result type.
+      SPV_BinaryOp<mnemonic, type, type,
+                   !listconcat(traits,
+                               [NoSideEffect, SameOperandsAndResultType])> {
+}
+
+class SPV_LogicalOp<string mnemonic, Type operandsType,
+                    list<OpTrait> traits = []> :
+      // Result type is SPV_Bool.
+      SPV_BinaryOp<mnemonic, SPV_Bool, operandsType,
+                   !listconcat(traits, [NoSideEffect, SameTypeOperands])> {
```

Comment:
We may also want `SameOperandsAndResultShape` here.

Reply:
Oh, thanks! Fixed.